### ISSUE TITLE
Investigation: Validation about 2d-array texture view upon 3D texture and vice versa, and more

### ISSUE THREAD

**Issue by Richard-Yunchao** [gpuweb/gpuweb#1372]
[Texture dimension types](https://gpuweb.github.io/gpuweb/#enumdef-gputexturedimension)  and [texture view dimension types](https://gpuweb.github.io/gpuweb/#enumdef-gputextureviewdimension) are different. When we create a texture view with a specific dimension type, which texture dimension type is valid? For example, whether we can create a 2D Array texture view upon a 3D texture and vice versa? This investigation attempts to get a conclusion for this question. 

## Question ##
In WebGPU, texture dimensions include `1d`, `2d`, `3d`. If the dimension is `2d` and size.depth is greater than 1, it is actually a `2d-array` texture. So there are actually 4 texture dimension types. While texture view dimensions include `1d`, `2d`, `2d-array`, `cube`, `cube-array`, `3d`. Texture and TextureView dimension types are different. There is no obvious bijection between them. So the question is: when we create a texture view with a given dimension type, which texture dimension type is valid (Y or N)? I try to answer this question (by my institution) below:
TextureView Dimension Type | 1d Texture | 2d Texture | 2d-array Texture* | 3d Texture |
-------------| ------------------------| -------------------------|----------------------|-------------------------
1d TextureView | Y | N | N | N
2d TextureView | N | Y | Y? | Y?
2d-array TextureView | N | Y? | Y | ? |
cube TextureView | N | N | Y* | ?
cube-array TextureView | N | N | Y* | ?
3d TextureView | N | N | ? | Y 

Notes:
* cube and cube-array TextureViews can be created upon 2d-array textures with a few more requirements like width equals to height, and its corresponding texture's array layers should be greater than 6 or multiple of 6, and so on. So I marked `Y*` for cube and cube-array TextureView upon 2d-array texture. 
* cube and cube-array TextureViews can be seen as special 2d-array texture views. So I marked `?` for 2d-array, cube, and cube-array TextureView upon 3d texture. I also marked `?` for 3d TextureView upon 2d-array texture. 

So the most important question is: whether we can create 2d-array texture view upon a 3D texture and vice versa. Another question is we need to make sure whether 2d view can be created upon 2d-array/3d texture and vice versa.    

### Vulkan ###
Vulkan exposes a flag named `VK_IMAGE_CREATE_2D_ARRAY_COMPATIBLE_BIT` via [VkImageCreateFlagBits](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VkImageCreateFlagBits) in [VkImageCreateInfo](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VkImageCreateInfo), in order to support 2d/2d-array texture view upon 3d texture, see `Table 15. Image and image view parameter compatibility requirements` under [Image Views](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#resources-image-views). 

This flag is supported in Vulkan 1.1 core feature, and Vulkan 1.0 with extension [VK_KHR_maintenance1](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VK_KHR_maintenance1)

However, there is no flag which can support 3D texture view compatible upon 2d-array texture. 
### Metal ###
The concept of texture view in Metal is really weak. I can't even find objects like `MTLTextureView` or `MTLTextureViewType` for texture view. The only one we can use is `MTLTexture` and `MTLTextureType`. If we want to reinterpret a texture type into another texture type for some reason like reinterpret texture formats, reuse partial subresource, etc., we can call [makeTextureView](https://developer.apple.com/documentation/metal/mtltexture/2928180-maketextureview), which will return a new texture with the dimension types, formats, etc., you want. 

On the description page of `makeTextureView`, a table describes the compatibility between the new texture dimension types and its original dimension types. Unfortunately, there is no way to create 2d-array texture (view) upon 3d texture or vice versa.  
### D3D12 ###
I didn't find any related flag in [D3D12_RESOURCE_FLAGS](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ne-d3d12-d3d12_resource_flags) of [D3D12_RESOURCE_DESC](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_resource_desc). 

We can also look into the definition about resource view. Take [D3D12_SHADER_RESOURCE_VIEW_DESC](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_shader_resource_view_desc) as an example:
```
typedef struct D3D12_SHADER_RESOURCE_VIEW_DESC {
  DXGI_FORMAT         Format;
  D3D12_SRV_DIMENSION ViewDimension;
  UINT                Shader4ComponentMapping;
  union {
    D3D12_BUFFER_SRV                            Buffer;
    D3D12_TEX1D_SRV                             Texture1D;
    D3D12_TEX1D_ARRAY_SRV                       Texture1DArray;
    D3D12_TEX2D_SRV                             Texture2D;
    D3D12_TEX2D_ARRAY_SRV                       Texture2DArray;
    D3D12_TEX2DMS_SRV                           Texture2DMS;
    D3D12_TEX2DMS_ARRAY_SRV                     Texture2DMSArray;
    D3D12_TEX3D_SRV                             Texture3D;
    D3D12_TEXCUBE_SRV                           TextureCube;
    D3D12_TEXCUBE_ARRAY_SRV                     TextureCubeArray;
    D3D12_RAYTRACING_ACCELERATION_STRUCTURE_SRV RaytracingAccelerationStructure;
  };
} D3D12_SHADER_RESOURCE_VIEW_DESC;
```
The explanation about `ViewDimension` is that `A D3D12_SRV_DIMENSION-typed value that specifies the resource type of the view. This type is the same as the resource type of the underlying resource. This member also determines which _SRV to use in the union below.`. This description of `ViewDimension` means that the resource view type should be exactly the same with its underlying resource. Moreover, definitions like [D3D12_TEX3D_SRV](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_tex3d_srv) and [D3D12_TEX2D_ARRAY_SRV](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_tex2d_array_srv) also assume the underlying resource is exactly a 3D texture and a 2d-array texture respectively.  

Other resource views like [D3D12_RENDER_TARGET_VIEW_DESC](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_render_target_view_desc) and [D3D12_UNORDERED_ACCESS_VIEW_DESC](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_unordered_access_view_desc) are basically the same, although their union numbers are less than SRV's and their definitions like D3D12_TEX3D_RTV (no mipLevelCount because all RTV should be the same size) and D3D12_TEX2D_ARRAY_UAV might differ slightly from D3D12_TEX3D_SRV and D3D12_TEX2D_ARRAY_SRV.   
## Conclusion ## 
* Now that we can't create 2d array texture view upon 3d texture and vice versa in Metal and D3D12, I think we should conclude that 2d-array (and cube/cube-array) texture view upon 3d texture and vice versa should be invalid in WebGPU. 
* We can't create 2d texture view upon 3d texture in Metal and D3D12 (it is allowed in Vulkan with a flag), although we can create a 2d texture view upon 3d texture in theory (via specify the mip level and depth of a given 3d texture). Likewise, we can't create 2d texture view upon a 2d-array texture in D3D12 (It is allowed in Vulkan and Metal). We can look into [D3D12_TEX2D_SRV](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_tex2d_srv). There is no parameter to set a particular arrayLayer, which means that it assume the underlying resource is a 2d texture, not a 2d-array texture. So we should conclude that 2d texture view upon 3d or 2d-array texture is not valid.
* We can create 2d-array texture view upon a 2d texture, because we can create 2d texture view upon a texture texture. And a 2d texture view can be seen as one-layer 2d-array texture view. But such a 2d-array texture view is useless. So 2d-array texture view upon 2d texture can be valid or invalid. I am neutral on this. 

So my proposal about texture view dimension type compatibility with texture dimension type is listed as follows:
TextureView Dimension Type | 1d Texture | 2d Texture | 2d-array Texture* | 3d Texture |
-------------| ------------------------| -------------------------|----------------------|-------------------------
1d TextureView | Y | N | N | N
2d TextureView | N | Y | N | N
2d-array TextureView | N | Y/N | Y | N |
cube TextureView | N | N | Y* | N
cube-array TextureView | N | N | Y* | N
3d TextureView | N | N | N | Y

**Comment by Kangz** [comment]
FYI I filed https://github.com/KhronosGroup/Vulkan-Docs/issues/1438 to clarify the Vulkan spec for some validation rule I had trouble finding.

### Metal

> The concept of texture view in Metal is really weak. I can't even find objects like MTLTextureView or MTLTextureViewType for texture view.

Metal is a bit different from other APIs, you can basically see each `MTLTexture` as a combination of a `(Ref<TextureInternal>, unique_ptr<TextureViewInternal>`. `makeTextureView` creates a new `MTLTexture` that contains a ref to the same `TextureInternal` but a new `TextureViewInternal`: no copy of the texture content happens.

Note that Metal does allow rendering to a slice of a 3D texture as described in the [documentation for MTLRenderPassDescriptor](https://developer.apple.com/documentation/metal/mtlrenderpassattachmentdescriptor):

> You must set the attachment’s texture property. The level, slice, and depthPlane properties specify the mipmap level, slice, and depth plane (for 3D textures) of the texture, respectively.

### D3D12

In D3D12 you can use a 2D array SRV as a non-array 2D texture in the shader. Likewise for cube vs. cube array. We verify this experimentally in Dawn because we always use `TEX2DARRAY` and `CUBEARRAY` even for texture view dimension `"2d"`.

### Discussion for the conclusion

D3D12 allows reinterpreting a 2D array texture as a 2D view (just set the layer count to 1 and use it as a 2D texture in the shader, same idea for RTV DSV and UAV). So I think we should allow reinterpreting 2D -> 2D array. Likewise, because WebGPU has strict rules on the compatibility between the viewDimension and the shader declaration, we should allow 2D array -> 2D reinterpretation.

At least Vulkan and Metal allow rendering to a slice of a 3D textures, although I'm not 100% sure for D3D12. This sounds useful but would likely require special flags on texture creation and additional logic in validation surrounding texture views. We could leave this out of MVP and revisit later.

**Comment by kvark** [comment]
@Richard-Yunchao thank you for investigating this! It's a nasty corner case that needed to be clarified for good.

@Kangz nicely expressed what I wanted to say, so we are in agreement there. Casting 2D -> 2D array view is useful to make shaders more generic: a shader expecting 2D array view can be made to work with a 2D texture this way.

And as for the array rendering - let's leave it out of MVP. The shader side of this problem is more complicated that the host side, and we don't need to dive into that just yet.

**Comment by kainino0x** [comment]
> So I think we should allow reinterpreting 2D -> 2D array.

We don't even have 2D non-array textures though, only 2D array textures and 2d/2d-array texture views.

**Comment by kainino0x** [comment]
> We don't even have 2D non-array textures though, only 2D array textures and 2d/2d-array texture views.

Unless we have special capabilities for depth=1 2d textures. I don't think we do?

**Comment by Richard-Yunchao** [comment]
> > So I think we should allow reinterpreting 2D -> 2D array.
> 
> We don't even have 2D non-array textures though, only 2D array textures and 2d/2d-array texture views.

**Comment by Richard-Yunchao** [comment]
> > We don't even have 2D non-array textures though, only 2D array textures and 2d/2d-array texture views.
> 
> Unless we have special capabilities for depth=1 2d textures. I don't think we do?

Literally, we only have 2d texture in WebGPU, not 2d-array texture: https://gpuweb.github.io/gpuweb/#enumdef-gputexturedimension. 

BTW, 2d texture can be seen as one layer 2d-array texture, it can also be seen as depth=1 3d texture. But that would be a little ambiguous. So I tend to say that we have 4 texture dimension types 1d, 2d, 2d-array, 3d in this investigation.

**Comment by Kangz** [comment]
(deleted an comment that was submitted too early)

> BTW, 2d texture can be seen as one layer 2d-array texture, it can also be seen as depth=1 3d texture. But that would be a little ambiguous. So I tend to say that we have 4 texture dimension types 1d, 2d, 2d-array, 3d in this investigation.

The problem is that underlying APIs don't allow reinterpreting a 2D texture as a 3D texture of depth=1. IMHO in WebGPU we should not mention 2D array as a separate dimensionality because that will confuse people even more. It's simpler to tell them that 2D textures are always arrayed but that the array size can be 1.

Given the comments above I think the table should be the following (with constraints on cubes of course):

TextureView Dimension Type | 1d Texture | 2d Texture | 3d Texture |
-------------| ------------------------| -------------------------|----------------------
1d TextureView | Y | N | N
2d TextureView | N | Y | N
2d-array TextureView | N | Y | N |
cube TextureView | N | Y | N
cube-array TextureView | N | Y | N
3d TextureView | N | N | Y

**Comment by Richard-Yunchao** [comment]
That's the same with my proposal if we don't introduce 2d-array texture. And it is much simpler. @Kangz. I guess we can workaround 2d view upon 2d-array texture and vice versa in D3D12. For example, [D3D12_TEX2D_SRV](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_tex2d_srv) doesn't have any parameter to specify a particular arrayLayer if the underlying texture is a 2d-array texture.  but we can use a D3D12_TEX2D_ARRAY_SRV with arrayLayerCount=1 for to implement 2d view in D3D12 and wordaround this issue. For 2d-array view upon 2d texture, it is similar.

**Comment by Richard-Yunchao** [comment]
> Note that Metal does allow rendering to a slice of a 3D texture as described in the [documentation for MTLRenderPassDescriptor](https://developer.apple.com/documentation/metal/mtlrenderpassattachmentdescriptor):
> 
> > You must set the attachment’s texture property. The level, slice, and depthPlane properties specify the mipmap level, slice, and depth plane (for 3D textures) of the texture, respectively.

A good finding. Metal allows users to render into a slice of 3d texture as a 2d view. But I didn't see any other way to sample or write into a 3d texture as 2d view. @litherum and @grorg , any comments on this? 

If this is true, when we introduce 2d view upon 3d texture in future, the usage is limited to `RENDER_ATTACHMENT`. Other usages like `COPY_SRC`, `COPY_DST`, `SAMPLED`, `STORAGE` may not be allowed for such a 2d view (upon a 3d texture).

**Comment by Richard-Yunchao** [comment]
> The problem is that underlying APIs don't allow reinterpreting a 2D texture as a 3D texture of depth=1. IMHO in WebGPU we should not mention 2D array as a separate dimensionality because that will confuse people even more. It's simpler to tell them that 2D textures are always arrayed but that the array size can be 1.

Sounds like we should expose 2d-array texture type, not 2d texture type. :) 

Regarding the naming stuff, I feel the current names about TextureDescriptor and TextureViewDescriptor are weird and confusing. GPUTextureDescriptor uses Extent3D to describe its size. And Extent3D has 3 dimensions: width, height, depth. You see, there is no `array` or `layer` related stuff at all if we look into the names of Texture. 

TextureView is based on Texture, but there are names like baseArrayLayer and arrayLayerCount. So, what are these names if we don't have `arrayLayer` in TextureDescritptor at all? We are actually assuming that developers know 2d-array texture!

I personally think renaming size.depth to size.depthOrArrayLayers in TextureDescriptor (it is like D3D12, if we don't want to introduce a separate parameter for arrayLayers like Vulkan/Metal) makes more sense, in order to be consistent with TextureViewDescirptor. 

Well, adding some notes/explanations like `2D textures are always arrayed but that the array size can be 1` can alleviate the confusing.

**Comment by Kangz** [comment]
Agreed that conceptually all 2D textures are 2D array. It's probably worth a note about this somewhere. We talked at length (at least two times) about the naming of merging of depth/arraySize and its naming in the group. Overall we reached the conclusion to use just `depth`.

I'm not sure what to do for the `GPUTextureViewDescriptor`, it's probably fine to keep the current names (but why is it `baseXXX` and not `firstXXX`?) if there's a clear note that explains: "GPUTextureDimension "2d" means the texture is an array of layers which are all 2D textures and the array depth is defined by .size.depth. When creating a texture view an interval of the array layers can be chosen with firstArrayLayer and arrayLayerCount".

**Comment by kainino0x** [comment]
> Sounds like we should expose 2d-array texture type, not 2d texture type. :)

That's correct, we simply don't make a distinction as the distinction is not meaningful until you get to texture views.

> I personally think renaming size.depth to size.depthOrArrayLayers in TextureDescriptor (it is like D3D12, if we don't want to introduce a separate parameter for arrayLayers like Vulkan/Metal) makes more sense, in order to be consistent with TextureViewDescirptor.

We considered depthOrArrayLayers before here: https://github.com/gpuweb/gpuweb/pull/1081#issuecomment-696470308
I am still personally in favor of the name. Especially because users don't actually _have_ to type it as we allow size to be specified as an array of [width, height, depthOrArrayLayers].

**Comment by Kangz** [comment]
I'm not sure if the result of these discussions have been encoded in the spec yet.
